rdata/unparser/_unparser.py

```diff
@@ -0,0 +1,170 @@
+    def unparse_int(self, value: int | np.int32) -> None:
+        """Unparse an integer value."""
+        self._unparse_array_values(np.array([value]))
+
+    def unparse_array(self, array: npt.NDArray[Any]) -> None:
+        """Unparse an array of values."""
+        # Expect only 1D arrays here
+        assert array.ndim == 1
+        self.unparse_int(array.size)
+        self._unparse_array_values(array)
+
+    @abc.abstractmethod
+    def _unparse_array_values(self, array: npt.NDArray[Any]) -> None:
+        """Unparse the values of an array."""
+
+    @abc.abstractmethod
+    def unparse_string(self, value: bytes) -> None:
+        """Unparse a string."""
+
+    def unparse_r_data(self, r_data: RData, *, rds: bool = True) -> None:
+        """Unparse an RData object."""
+        self.unparse_magic(None if rds else r_data.versions.format)
+        self.unparse_header(r_data.versions, r_data.extra)
+        self.unparse_r_object(r_data.object)
+
+    def unparse_r_object(self, obj: RObject) -> None:  # noqa: C901, PLR0912
+        """Unparse an RObject object."""
+        # Some types include attributes and tag with data while some add them
+        # later. These booleans keep track of whether attributes or tag
+        # has been done already
+        attributes_done = False
+        tag_done = False
+
+        # Unparse info bytes
+        info = obj.info
+        self.unparse_int(pack_r_object_info(info))
+
+        # Unparse data
+        value = obj.value
+        if info.type in {
+           RObjectType.NIL,
+           RObjectType.NILVALUE,
+        }:
+            # These types don't have any data
+            assert value is None
+
+        elif info.type == RObjectType.SYM:
+            self.unparse_r_object(value)
+
+        elif info.type in {
+            RObjectType.LIST,
+            RObjectType.LANG,
+            # Parser treats the following equal to LIST.
+            # Not tested if they work
+            # RObjectType.CLO,
+            # RObjectType.PROM,
+            # RObjectType.DOT,
+            # RObjectType.ATTRLANG,
+        }:
+            if info.attributes:
+                assert obj.attributes is not None
+                self.unparse_r_object(obj.attributes)
+                attributes_done = True
+
+            if info.tag:
+                assert obj.tag is not None
+                self.unparse_r_object(obj.tag)
+                tag_done = True
+
+            for element in value:
+                self.unparse_r_object(element)
+
+        elif info.type in {
+            RObjectType.CHAR,
+            RObjectType.BUILTIN,
+            # Parser treats the following equal to CHAR.
+            # Not tested if they work
+            # RObjectType.SPECIAL,
+        }:
+            self.unparse_string(value)
+
+        elif info.type in {
+            RObjectType.LGL,
+            RObjectType.INT,
+            RObjectType.REAL,
+            RObjectType.CPLX,
+        }:
+            self.unparse_array(value)
```

Comment:
Should we consider ND arrays? IIRC we just needed to add some attributes, and it may be very useful for some users.

Reply:
ND arrays work already. The original ND array is flattened to 1D and attributes are added in `convert_to_r_object()`. Then this unparser line sees only the flattened array.